src/analysis/processing/qgsalgorithmpackage.cpp

```diff
@@ -74,14 +75,197 @@ QgsPackageAlgorithm *QgsPackageAlgorithm::createInstance() const
+    if ( project && ! project->relationManager()->relations().isEmpty() )
+    {
+
+      // Infinite recursion should not happen in the real world but better safe than sorry
+      const int maxRecursion { 10 };
+      int recursionGuard { 0 };
+
+      // This function recursively finds referenced layers
+      const auto findReferenced = [ &project, &feedback, &recursionGuard, &layers, this ]( const QgsVectorLayer * vLayer, bool onlySaveSelected, auto &&findReferenced ) -> void
+      {
+        const QgsVectorLayer *originalLayer { qobject_cast<QgsVectorLayer *>( project->mapLayer( mClonedLayerIds.value( vLayer->id(), vLayer->id() ) ) ) };
+        Q_ASSERT( originalLayer );
+        const QList<QgsRelation> relations { project->relationManager()->referencingRelations( originalLayer ) };
+        for ( const QgsRelation &relation : std::as_const( relations ) )
+        {
+
+          QgsVectorLayer *referencedLayer { relation.referencedLayer() };
+          QgsVectorLayer *relatedLayer { nullptr };
+
+          // Check if the layer was already in the export list
+          bool alreadyAdded { false };
+          for ( const auto &layerToExport : std::as_const( mLayers ) )
+          {
+            const QString originalId { mClonedLayerIds.value( layerToExport->id() ) };
+            if ( originalId == referencedLayer->id() )
+            {
+              relatedLayer = qobject_cast<QgsVectorLayer *>( layerToExport.get() );
+              alreadyAdded = true;
+              break;
+            }
+          }
+
+          if ( !alreadyAdded )
+          {
+            feedback->pushInfo( QObject::tr( "Adding referenced layer '%1'" ).arg( referencedLayer->name() ) );
+            relatedLayer = referencedLayer->clone();
+            mLayers.emplace_back( relatedLayer );
+            mClonedLayerIds.insert( relatedLayer->id(), referencedLayer->id() );
+          }
+
+          // Export only relevant features unless the layer was in the original export list and no selection was made on that layer
+          // in that case the user explicitly marked the layer for export and it is supposed to be exported fully.
+          if ( onlySaveSelected )
+          {
+            if ( ! layers.contains( qobject_cast<QgsMapLayer *>( referencedLayer ) ) || referencedLayer->selectedFeatureCount() > 0 )
+            {
+              Q_ASSERT( relatedLayer );
+              QgsFeatureIds selected;
+              QgsFeatureIterator it { vLayer->getSelectedFeatures() };
+              QgsFeature selectedFeature;
+              while ( it.nextFeature( selectedFeature ) )
+              {
+                QgsFeature referencedFeature { relation.getReferencedFeature( selectedFeature ) };
+                if ( referencedFeature.isValid() )
+                {
+                  selected.insert( referencedFeature.id() );
+                }
+              }
+              relatedLayer->selectByIds( selected, Qgis::SelectBehavior::AddToSelection );
+            }
```

Comment:
This is expensive code to run in the main thread (blocking). Can you move it to the processAlgorithm call instead?